KEYCLOAK-7470 Ability to order client scopes

mposolda opened the pull request.

mposolda committed:
KEYCLOAK-7470 Ability to order client scopes

mposolda commented:
@pedroigor @hmlnarik @stianst Anyone has a chance to review?

hmlnarik commented on themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties:
```diff
@@ -827,6 +827,8 @@ client-scope.display-on-consent-screen=Display On Consent Screen
 client-scope.consent-screen-text=Consent Screen Text
 client-scope.consent-screen-text.tooltip=Text, which will be shown on consent screen when this client scope is added to some client with consent required. Defaults to name of client scope if it's not filled
+client-scope.gui-order=GUI order
+client-scope.gui-order.tooltip=Number defining order of the provider in GUI (eg. on Consent page).
```
Specify order of the provider in GUI (e.g. in Consent page) as integer

mposolda commented on services/src/main/java/org/keycloak/forms/account/freemarker/model/ApplicationsBean.java:
```diff
@@ -72,18 +76,17 @@ public ApplicationsBean(KeycloakSession session, RealmModel realm, UserModel use
                 }
             }
+            List<String> clientScopesGranted = orderedScopes.stream()
+                    .map(clientScope -> clientScope.getConsentScreenText())
```
Good to know. Thanks for pointing this.

mposolda commented:
@hmlnarik Addressed your feedback with the exception of the comparator. It can't return 0 unless objects are the same (equals). Because GUI Order is not mandatory field for the client scopes and identity providers, so  returning 0 would result that some object without GUIOrder specified, won't be put to the collection. Added one more unit test for this (IdentityProviderBeanTest.testIdentityProviderComparatorForEqualObjects)

hmlnarik commented:
Reason is that `TreeSet` is used which does not allow duplicates. List should be used instead

mposolda commented:
@hmlnarik I've changed everywhere to the List as we discussed
